Add documentation for moderated sessions (#9425)
gravitational/teleport: docs/pages/setup/reference/moderated-sessions.mdx

---
title: Moderated Sessions
description: Moderated Sessions RBAC Reference
h1: Moderated Sessions
---

# Moderated Sessions

> Moderated Sessions makes use of RBAC to control access to and within sessions.

Author: 👍

Here is a collection of details on configuring Moderated Sessions.

## RBAC

Moderated Sessions makes use of RBAC policies to allow for fine grained control over
who can join a session and who are required to be present to start one.

> The system is based around **require policies** and allow policies. Require policies define a set of conditions

Reviewer: ```suggestion
The system is based around **require policies** and **allow policies**. 

Require policies define a set of conditions
```

Since we introduce both concepts in the same sentence, I think it looks a bit cleaner to separate it from the sentence that goes in depth on require policies.

and under what conditions they may join a session.

> ## Configuring moderated sessions

Reviewer: My biggest issue with this section is that it is more like a guide than a reference. When reading a reference, I would expect a comprehensive set of available options (e.g., see our docs guide: https://goteleport.com/docs/docs/#reference-manuals). Instead, each section describes or illustrates a topic without being exhaustive.

I recommend reworking this section to have the following structure:
1. Add headers for `require_session_join` and `join_sessions`
1. Under each header, list the required/optional keys and their possible values (or the types of these values), as well as what these options configure. I suggest using a table.
1. Add a separate section explaining the filter syntax. The section should have the same heading level as the `require_session_join` and `join_sessions` sections. When we indicate the type of the `filter` option, we can link to that section.
1. Change "Session kinds" to use a table. We can link to this section when specifying the type of the `kinds` value.
1. Move the "Participant modes" section into the same heading level as the `join_sessions`/`require_session_join` sections. That way, when we indicate the type of the `modes` field, we can say that it's a list of participant modes, and link to that section.

As an illustration of what I mean, I think some parts of the Istio reference documentation do a good job at laying out tables of YAML config options and specifying the types of the values. In [this example](https://istio.io/latest/docs/reference/config/networking/gateway/#Gateway), we can see that the the `servers` field is a `Server[]`, and a separate section specifies the fields of `Server`.

> **Allow policies** are used to define what sessions a user can join

Reviewer: ```suggestion
Allow policies are used to define what sessions a user can join
```
Suggested removing the bold here and bolding the term when we first introduced it.

and under what conditions they may join a session.

> 

Reviewer: Since I've suggested using the current "RBAC" section intro as the intro for the whole page, let's add a new H2 heading here. Maybe "Configuring Moderated Sessions"? This will include the H3 headings re: specific config options.

```suggestion

## Configuring Moderated Sessions
```

### Session kinds

> Require and allow policies have to specify which sessions they apply to. Valid options are `ssh` and `k8s`.

Reviewer: Which config field would you use these values for?

Author: I've moved the section down and pointed to the example configs on how to specify the kinds.

### Filters

A filter expression or predicate is a string statement used to define logic based on a set of input variables.
These follow common programming language syntax and allow for more detailed control.
`user.name == "adam" || contains(user.traits, "view")` is a simple example of a filter expression
that evaluates to true if the user is Adam or if the user has the trait `view`.

Require policies can specify which users they consider as valid with a filter expression. They work like
other filters used within Teleport such with access requests. The filter context has a `user` object defined

> which the set fields `roles`, `traits` and `name`.

Reviewer: While I know a full example is coming up in this page, I think other readers may be expecting an example here. Otherwise, it may not be clear where a filter appears within a config resource.

Author: Do you mean just a filter example or like a minimal policy with a filter? The reason I am hesitant is that it'll be pretty much the exact same as the examples below since there aren't really optional fields on a policy.

If a user has any attached V5 roles, the new RBAC access checks will be enforced.

### Example require policy

The policy below specifies that a given role
must have a minimum of two users with the role `auditor` and the mode `moderator` present in the session
to start it. The policy applies to SSH and Kubernetes sessions only.

When a user with this require policy starts a session, it will be pending
until the policy is fulfilled.

```yaml
kind: role
metadata:
  name: prod-access
spec:
  allow:
    require_session_join:
      - name: Auditor oversight
        filter: 'contains(user.roles, "auditor")'
        kinds: ['k8s', 'ssh']
        modes: ['moderator']
```

### Example allow policy

The following allow policy attaches to the role `auditor` and allows one to join
SSH and Kubernetes sessions started by a user with the role `prod_access` as a moderator or observer.

```yaml
kind: role
metadata:
  name: auditor
spec:
  allow:
    join_sessions:
      - name: Auditor oversight
        roles : ['prod-access']
        kinds: ['k8s', 'ssh']
        modes: ['moderator', 'observer']
```

## MFA-based presence

> When `mfa_required` is set to `true` via role or cluster settings, Teleport enforces

Reviewer: - If MFA-based presence is another RBAC config setting, should we include it in the previous section (i.e., as an H3 heading) rather than as a new H2-level section?
- I think it would be helpful to add examples of what `mfa_required` looks like in role or cluster settings. This would make it clear where in a config spec we would expect to find this field (e.g., directly under `spec` or under another field).

Author: I got the MFA setting name wrong. We've got a page on per session MFA which is now linked.

MFA-based presence checks for moderators.
This requires that all moderators wishing to join have a configured U2F or WebAuthn MFA token.

Every 30 seconds, Teleport will issue a prompt to the user in the terminal, asking them
to press their MFA token in the next 15 seconds. This will happen continously during the session
and exists so that moderators are always present and watching a given session.

If no MFA input is received within 60 seconds, the user is kicked
from the session which may pause it, if RBAC policies are no longer met.

> ### Participant Modes

Reviewer: Should we make this an H2-level section since it doesn't have to do with configuring RBAC resources?

Author: I've liberally decided to convert most of the headers to H2, maybe It'd make sense to group Session kinds, Filters, example policies and MFA-based presence under a H2 named RBAC configuration and make them H3, what do you think?

Author: I've reworked headings now, take a look

A participant joining a session will always have one of three modes:

> - Peers can join and collaborate in a session. They can view output and send input.

Reviewer: ```suggestion
- **Peers** can join and collaborate in a session. They can view output and send input.
```

> - Moderators can join and watch a session. They can view output and forcefully terminate the session at will.

Reviewer: ```suggestion
- **Moderators** can join and watch a session. They can view output and forcefully terminate the session at will.
```

> - Observers can join and watch a session. They cannot control the session in any way.

Reviewer: ```suggestion
- **Observers** can join and watch a session. They cannot control the session in any way.
```

> When joining a session with `tsh join` or `tsh kube join`, a mode can be specified with `--mode <mode>`

Reviewer: ```suggestion
When joining a session with `tsh join` or `tsh kube join`, a user can specify a mode with the `--mode <mode>` flag
```
I think this sounds more direct in the active voice.

> where mode is one of `peer`, `moderator` or `observer`. By default, the mode is `peer` for SSH and

Reviewer: ```suggestion
, where the mode is one of `peer`, `moderator` or `observer`. By default, the mode is `peer` for SSH and
```
